[TIL-125] 기술 학습 문제 풀이 구현

dongyeon-0822 opened the pull request:
## 이슈 번호(링크)

[TIL-125](https://soma-til.atlassian.net/browse/TIL-125)

<br>

## 개요

- 기술 학습 문제 풀이 구현

<br>

## 내용

- UserProblem 도메인 추가
- 점수 계산(채점) 로직: 점수 랜덤으로 결정 -> 상태 결정
- 성공 및 실페 에러 코드 추가
- UserProblemServiceTest 작성

<br>

## 리뷰어한테 할 말

고민했던 부분들🫠 
아래 내용들 모두 고민스러웠던 부분이기 때문에 더 좋은 의견 있으시면 남겨주세요:) 

- 전체적인 구조: domain과 Api에서는 Problem 파일에 합치고 application(service)단에서는 나중에 더 로직이 많아질 거라 생각해서 UserProblem을 분리했습니다. 
- Controller 관련: answer 입력에 대한 검증은 프론트에서 검증하겠지만, 백엔드에서도 필요하다고 생각하여 UserProblemRequest에서 `@NotBlank`로 처리했습니다. 
- Controller 관련: Response에 우선 status와 solution만 반환하는데, 추가로 보낼 값이 있다면 추가하겠습니다. 
- Service 관련: ProblemStatus enum을 작성할 때, default 값은 UnSolved가 필요하다고 생각했고, 이에 대한 defalut 처리를 어디에서 해야할지 고민했는데, 우선 dto에서 entity를 생성할 때, 아래와 같이 했습니다. 
```java
public UserProblem toEntity(User user, Problem problem) {
        return UserProblem.builder()
            .user(user)
            .problem(problem)
            .answer(answer)
            .status(ProblemStatus.UNSOLVED)
            .build();
    }
```
- Service 관련: 아래 코드에서 dto에서 entity를 만든 후에, 채점이 되기 때문에 entity에 새로운 score와 status에 대한 set이 필요하여, entity 내부에 score, status만 set하는 로직을 추가했습니다. 
```java 
@Transactional
    public UserProblemResultDto solveProblem(UserProblemDto userProblemDto) {
        User user = userRepository.getById(userProblemDto.userId());
        Problem problem = problemRepository.getById(userProblemDto.problemId());
        UserProblem userProblem = userProblemDto.toEntity(user, problem);

        int score = calculateScore();
        ProblemStatus status = determineStatus(score);
        userProblem.setResultProblem(score, status);
        UserProblem userProblemResult = userProblemRepository.saveUserProblem(userProblem);

        return UserProblemResultDto.of(userProblemResult);
    }
```
- Entity 관련: save에도 검증 로직을 구현하고자 했는데, 이 부분도 find처럼 repository에 구현하는게 맞을지, 아예 검증을 하지 않는게 맞을지 모르겠습니다. 
```java 
public interface UserProblemRepository extends JpaRepository<UserProblem, Long> {

    default UserProblem saveUserProblem(UserProblem userProblem) {
        try {
            return save(userProblem);
        } catch (Exception e) {
            throw new BaseException(ProblemErrorCode.FAILED_TO_SAVE_PROBLEM);
        }
    }
}
```
- Success Response
<img width="711" alt="image" src="https://github.com/user-attachments/assets/f7a77cf8-5808-47c9-aba3-7133a60bd51b">


[TIL-125]: https://soma-til.atlassian.net/browse/TIL-125?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

dongyeon-0822 committed:
feat: 기술 학습 문제 풀이 구현

- UserProblem 도메인 추가
- 점수 계산(채점) 로직 랜덤하게 구현
- 성공 및 실페 에러 코드 추가
- UserProblemServiceTest 작성

CodeDiary18 commented on til-domain/src/main/java/com/til/application/userProblem/UserProblemService.java:
```diff
@@ -0,0 +1,64 @@
+@Service
+@RequiredArgsConstructor
+@Transactional(readOnly = true)
+public class UserProblemService {
+
+    private final UserProblemRepository userProblemRepository;
+
+    private final ProblemRepository problemRepository;
+
+    private final UserRepository userRepository;
```
P4: UserProblem~ 대신 ProblemSolve~ 같은 네이밍도 괜찮을 것 같습니다!

dongyeon-0822 replied:
원래 Entitiy, Service, Controller 파일명은 UserProblem~ 으로 하고 그 외 DTO들은 SolveProblem~ 으로 했었는데, 이 부분도 고민하다 우선 UserProblem~으로 했습니다!
저도 SolveProblem이나 ProblemSolve이 더 적합하다고 생각하는데, Entitiy, Service, Controller 파일명부터 모두 변경하는게 좋을까요?

CodeDiary18 replied:
UserProblem으로 작업된 부분이 있다면 일관성있게 변경하면 좋을 것 같습니다!

CodeDiary18 commented on til-domain/src/main/java/com/til/application/userProblem/UserProblemService.java:
```diff
@@ -0,0 +1,64 @@
+        return random.nextInt(101);
+    }
+
+    private ProblemStatus determineStatus(int score) {
+        if (score >= 90) {
+            return ProblemStatus.PERFECT;
+        } else if (score >= 75) {
+            return ProblemStatus.GOOD;
+        } else if (score >= 50) {
+            return ProblemStatus.NORMAL;
+        } else if (score >= 25) {
+            return ProblemStatus.BAD;
+        } else {
+            return ProblemStatus.WRONG;
+        }
+    }
```
P3: 90, 75, 50, 25 부분도 채점 기준 같은걸로 분리되면 좋을 것 같습니다!  
어딘가 아래와 같이 정의되어 있다고 할 때
```java
@Getter
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
class enum ScoreRange {
    PERFECT(90, ProblemStatus.PERFECT),
    GOOD(75, ProblemStatus.GOOD),
    NORMAL(50, ProblemStatus.NORMAL),
    BAD(25, ProblemStatus.BAD);

    private int criteriaScore;
    private ProblemStatus status;
}
```

아래와 같이 변경해볼 수 있을 것 같습니다
```suggestion
    private ProblemStatus determineStatus(int score) {
      return Arrays.stream(ScoreRange.values())
                      .filter(range -> score >= range.getScore())
                      .map(ScoreRange::getStatus).findFirst()
                      .orElse(ProblemStatus.WRONG);
    }
```

CodeDiary18 commented on til-domain/src/main/java/com/til/domain/problem/dto/UserProblemDto.java:
```diff
@@ -0,0 +1,25 @@
+                             String answer
+) {
+
+    public UserProblem toEntity(User user, Problem problem) {
+        return UserProblem.builder()
+            .user(user)
+            .problem(problem)
+            .answer(answer)
+            .status(ProblemStatus.UNSOLVED)
```
P3: 답변이 제출되었을 때가 UserProblem이 생성되는 거라 UNSOLVED 보다는 추후 비동기 방식의 채점을 고려해서 채점(?) 대기 상태를 의미하는 `PENDING`을 기본값으로 설정해도 좋을 것 같습니다!

CodeDiary18 commented on til-domain/src/main/java/com/til/domain/problem/repository/UserProblemRepository.java:
```diff
@@ -0,0 +1,18 @@
+
+public interface UserProblemRepository extends JpaRepository<UserProblem, Long> {
+
+    default UserProblem saveUserProblem(UserProblem userProblem) {
+        try {
+            return save(userProblem);
+        } catch (Exception e) {
+            throw new BaseException(ProblemErrorCode.FAILED_TO_SAVE_PROBLEM);
+        }
+    }
```
P3: 사용자와 제출하고자 하는 문제의 id가 존재하는지 검증이 이뤄지고 난 뒤 save를 하는게 맞는거 같아서 save 단계에서 오류가 난다면 그냥 DB 자체의 문제로 INTERNAL_SERVER_ERROR로 처리해도 괜찮을 것 같습니다!  
정리하자면 save는 별도 커스텀으로 에러 핸들링이 필요 없을 것 같다는 의견입니다!

CodeDiary18 reviewed (commented):
작업하시면서 고민이 깊으셨을 것 같습니다🥲  

FK 관련해 종속성과 연관관계가 많아지는 것 같다는 생각이 듭니다(해당 PR에 대해서만 말씀드리는게 아니라 제가 진행중인 작업에서도 고민하고 있는 부분입니다..)  
사용자나 문제 이런식으로 분리할 수 있는 도메인에 대해서는 @JoinColumn의 foreignkey 옵션을 NO_CONSTRAINT로 설정해보는 방법도 고려해보면 좋을 것 같습니다!

doburn commented on til-domain/src/main/java/com/til/domain/problem/model/Problem.java:
```diff
@@ -53,4 +53,7 @@ public class Problem extends BaseTimeEntity {
 
     @OneToMany(mappedBy = "problem", cascade = CascadeType.ALL, orphanRemoval = true)
     Set<ProblemCategory> problemCategorySet = new HashSet<>();
+
+    @OneToMany(mappedBy = "problem", cascade = CascadeType.ALL, orphanRemoval = true)
+    Set<UserProblem> userProblemSet = new HashSet<>();
```
P3: 연관 관계에서 `Set` `List` 자료구조에 대해서도 논의해보면 좋을 것 같습니다!

dongyeon-0822 committed:
refactor: 기술 학습 문제 풀이 수정

- UserProblem -> SolveProblem (Repository, Entity 제외)
- score 제거 및 채점 로직 수정
- save 검증 제거, 테스트 제거
